include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1124,405 @@ AlignedVector<T>::fill(const T &value)
+          else
+            *this = Utilities::MPI::broadcast(
+              shmem_roots_communicator,
+              *this,
+              Utilities::MPI::n_mpi_processes(shmem_roots_communicator) - 1);
+        }
+    }
+  // We no longer need the shmem roots communicator, so get rid of it
+  {
+    const int ierr = MPI_Comm_free(&shmem_roots_communicator);
+    AssertThrowMPI(ierr);
+  }
+
+
+  // **** Step 3 ****
+  // At this point, all shmem groups have one shmem root process that has
+  // a copy of the data. This is the point where each shmem group should
+  // establish a shmem area to put the data into. As mentioned above,
+  // we know that the shmem roots are the last rank in their respective
+  // shmem_group_communicator.
+  //
+  // The process for all of this works as follows: While all processes in
+  // the shmem group participate in the generation of the shmem memory window,
+  // only the shmem root actually allocates any memory -- the rest just
+  // allocate zero bytes of their own. We allocate space for exactly
+  // size() elements (computed on the shmem_root that already has the data)
+  // and add however many bytes are necessary so that we know that we can align
+  // things to 64-byte boundaries.
+  //
+  // Before MPI 4.0, there was no way to specify that we want memory aligned to
+  // a certain number of bytes. This is going to come back to bite us further
+  // down below when we try to get a properly aligned pointer to our memory
+  // region, see the commentary there. Starting with MPI 4.0, one can set a
+  // flag in an MPI_Info structure that requests a desired alignment, so we do
+  // this for forward compatibility; MPI implementations ignore flags they don't
+  // know anything about, and so setting this flag is backward compatible also
+  // to older MPI versions.
+  //
+  // There is one final piece we can already take care of here. At the beginning
+  // of all of this, only the shmem_root knows how many elements there are in
+  // the array. But at the end of it, all processes of course need to know. We
+  // could put this information somewhere into the shmem area, along with the
+  // other data, but that seems clumsy. It turns out that when calling
+  // MPI_Win_allocate_shared, we are asked for the value of a parameter called
+  // 'disp_unit' whose meaning is difficult to determine from the MPI
+  // documentation, and that we do not actually need. So we "abuse" it a bit: On
+  // the shmem root, we put the array size into it. Later on, the remaining
+  // processes can query the shmem root's value of 'disp_unit', and so will be
+  // able to learn about the array size that way.
+  MPI_Win        shmem_window;
+  void *         base_ptr;
+  const MPI_Aint align_by   = 64;
+  const MPI_Aint alloc_size = Utilities::MPI::broadcast(
+    shmem_group_communicator,
+    ((size() * sizeof(T) + align_by - 1) / sizeof(T)) * sizeof(T),
```

Comment:
Can you comment on this choice some more? We need at least `size()*sizeof(T)`. If want a multiple of `align_by` shouldn't this be
```
((size() * sizeof(T) + align_by - 1) / align_by * align_by,
```
?

Reply:
Yes, you are right. I copied this line from @peterrum somewhere without thinking about it.

Reply:
Thanks for pointing it out -- I should really have thought about this more!

Reply:
Well, actually still not. This actually fails right away. What your formula does is find the smallest multiple of `align_by` bytes so that we can fit `size()*sizeof(T)` bytes into it. But that's not what we are after. What we need is this: What is the smallest size of a buffer (however that may be aligned) so that we are guaranteed that we can put `size()*sizeof(T)` bytes onto an aligned address?

The answer is `(size() * sizeof(T) + align_by - 1)`, which happens if the memory system gives us a pointer to an address that is one byte past an alignment boundary. In that case, we need to waste the first `align_by-1` bytes.

Reply:
That sounds reasonable. :slightly_smiling_face: